lib/withOnyx.js

```diff
@@ -101,47 +116,86 @@ export default function (mapOnyxToState) {
+             * This is a design flaw in Onyx itself as dispatching updates before initial hydration is not a correct event flow.
+             * We however need to workaround this issue in the HOC. The addition of initialValue makes things even more complex,
+             * since you cannot be really sure if the component has been updated before or after the initial hydration. Therefore if
+             * initialValue is there, we just check if the update is different than that and then try to handle it as best as we can.
+             *
              * @param {String} statePropertyName
              * @param {*} val
              */
             setWithOnyxState(statePropertyName, val) {
-                // We might have loaded the values for the onyx keys/mappings already from the cache.
-                // In case we were able to load all the values upfront, the loading state will be false.
-                // However, Onyx.js will always call setWithOnyxState, as it doesn't know that this implementation
-                // already loaded the values from cache. Thus we have to check whether the value has changed
-                // before we set the state to prevent unnecessary renders.
                 const prevValue = this.state[statePropertyName];
-                if (!this.state.loading && prevValue === val) {
-                    return;
-                }
 
+                // If the component is not loading (read "mounting"), then we can just update the state
                 if (!this.state.loading) {
-                    this.setState({[statePropertyName]: val});
+                    // Performance optimization, do not trigger update with same values
+                    if (prevValue === val || utils.areObjectsEmpty(prevValue, val)) {
+                        return;
+                    }
+
+                    this.setStateProxy({[statePropertyName]: val});
                     return;
                 }
 
                 this.tempState[statePropertyName] = val;
 
-                // All state keys should exist and at least have a value of null
-                if (_.some(requiredKeysForInit, key => _.isUndefined(this.tempState[key]))) {
+                // If some key does not have a value yet, do not update the state yet
+                const tempStateIsMissingKey = _.some(requiredKeysForInit, key => _.isUndefined(this.tempState[key]));
+                if (tempStateIsMissingKey) {
                     return;
                 }
 
-                // Leave untouched previous state to avoid data loss during pre-load updates.
-                // This handles case when setState was called before the setWithOnyxState.
-                // For example, when an Onyx property was updated by keyChanged before the call of the setWithOnyxState.
+                const stateUpdate = {...this.tempState};
+                delete this.tempState;
+
+                // Full of hacky workarounds to prevent the race condition described above.
```

Comment:
NAB / FWIW, I am looking at this code and it looks reasonable since the edge case is well described and this clearly shows the few cases we will run into. Maybe I am a hacker though 😄 Nice work either way.